### DISCUSSION TITLE
Destroying non-used entities

### DISCUSSION THREAD

**Discussion by menganha** [skypjack/entt#1153] [Q&A] [answered]
Hello All, 

I have the following general (opinion-based perhaps?) question: What's do you think is best approach to delete entities not being used from the registry?

The options I see:

1) Define an event: You trigger an event with the entity as the message and clean up all of them at a later point in the execution of the program. This is the approach I currently use. However it has the disadvantage of not keeping track of duplicated entities to be destroyed, and therefore one may destroy them more than once which leads to UB. Is there a way of checking in EnTT if an entity exists/is being used by the registry? I haven't found any method that does that but it would solve the issue by checking the existence before destroying it.

2) Add a dummy component, and clean the ones with this component at the same time at a later point in the execution of the program. This solves the issue of duplication as we would add twice the component but this is doesn't cause any issues.

3) Keep some sort of vector of entities to delete. At the point of deletion check for duplicates and delete the unique identifiers. Issue with this approach is this vector will be all over the place in the code and IMO it makes everything quite messy.

4) Remove all components of entities to be destroyed and then, at a later point, delete the set of these "orphaned" entities (very similar to that of 2) but pehaps more efficient?)

Thanks in Advanced,
Menganha

**Comment by skypjack** [comment, marked as answer]
My personal choice is for the second option, that is, marking entities with a dedicated component. By far better than anything else imho.
As a bonus point, it also has an optimized path if you pass the view iterators to the destroy function directly. So... 🙂

**Comment by menganha** [comment]
Hey, thanks skypjack for your answer. 

Regarding the issue with checking before hand if a entity is being used in the registry before attempting to destroy it: Is there any function on entt that one could use to do so?

**Reply by skypjack** [reply]
Do you mean to know if it's a valid entity? I think `registry.valid(entity)` is what you're looking for then?

**Reply by menganha** [reply]
Yes, exactly, thanks.
Would in general the approach of checkingit the validity before deleting them be more or less performant?

**Reply by skypjack** [reply]
Check for validity only if you're in doubts. As with anything else, an additional operation has an additional cost compared to not doing it. 🤷‍♂️ 
For example, if the entity is returned by a view that you're iterating, there is absolutely no reason in checking for validity. On the other hand, if you store entities aside in a vector and elaborate them later on for some reasons, checking could help to avoid subtle bugs.
It really depends on the use case though.

**Reply by menganha** [reply]
Got it. Thanks for your input.
